### DISCUSSION TITLE
ImportError: dlopen(..., 0x0002): symbol not found in flat namespace _PyCMethod_New

### DISCUSSION THREAD

**Discussion by ghyatzo** [PyO3/pyo3#3376] [Questions] [answered]
Hello, 

I am having trouble manually building a python package. I developed it without issues using a virtual env and `maturin develop`, as in: it works and can be imported without issues.
Now, this library is the backbone of a python plugin and I need to vendor the library with the plugin itself.

I've followed the wiki about manual building thoroughly (at least i think) and I can `cargo build` without problems.
After copying the `.dylib` file into a properly renamed `.so` file I get the following error when importing:
```
>>> import my_lib_package
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: dlopen(/correct/path/to/my/library/package/my_library_package.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace '_PyCMethod_New'
```
Incidentally the very first line of my `#[pymodule]` function is a `m.add_function` call, If i comment it out the second line is a `m.add_class::<>()` call and the symbol not found becomes `_PyObject_CallNoArgs`.

I am a bit at a loss on what to do, has anybody have any idea what might be the issue or what I might be doing wrong?

Here is some context (if you need anything else let me know):
My `Cargo.toml`:
```
[package]
name = "MyLib"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[lib]
name = "my_lib"
crate-type = ["cdylib"]

[dependencies]
mylib_in_rust = { path = "./ext/my_lib_in_rust" }
pyo3 = { version = "0.19", features = ["extension-module"] }
pyo3-asyncio = { version = "0.19", features = ["tokio-runtime"] }
tokio = "1.29.1"

[build-dependencies]
pyo3-build-config = "0.19.2"
```

**Machine**: M1 mac, Mac Os Ventura 13.1
**Other Info**: pyenv with only two versions: `3.8.10` (built with `--enable-shared`) and the `system` one.

**Comment by davidhewitt** [comment, marked as answer]
`PyCMethod_New` was added in Python 3.9 so you should check your build environment (you can do `PYO3_PRINT_CONFIG=1 cargo build` to get PyO3 to dump what it has detected). PyO3 has presumably detected a 3.9 interpreter (or newer) and is building with that one.

This is why it fails when you're importing on a 3.8 interpreter.

**Comment by ghyatzo** [comment]
Thanks a lot, I wrongly assumed that if building in a folder with a `.python-version` file pyo3 would use the local version set by `pyenv`.
Now with `PYO3_PYTHON="$(pyenv which python)" cargo build` it compiles and import correctly.

Strangely now, the lib works well when built with `maturin` while it acts up if manually built.
But that is probably on my end.

**Comment by TheGreekGod007** [comment]
dlopen(/Users/munzur/miniforge3/lib/python3.9/site-packages/scipy/spatial/_distance_wrap.cpython-39-darwin.so, 0x0002): symbol not found in flat namespace '_npy_copysign'
